### PR TITLE
ipasudorule: Add support for batch mode and multiple sudorules

### PR DESCRIPTION
**Pull request by rjeffman**
Currently, ipasudorule must add or modify a single sudorule at a time, incurring in more load in the server if there are many rules to be processed.

This patch adds suport for adding multiple sudorules in one playbook task by using the parameter 'sudorules' and defining a list of sudorules configurations to be ensured.

As multiple sudorules will be processed, the patch also enables batch mode processing of sudorules, trying to reduce the load on the server.

To be able to process parameter values in the 'sudorules' dictionary, the method to retrieve parameters from the module in lowercase was refactored, allowing any parameter value, either from the module or from a dict, to be converted.

Test 'tests/sudorule/test_sudorule_client_context.yml' was modified to include tasks with 'sudorules' to be executed both on the server or on the client context.

New tests were added to the sudorule test suite:

    tests/sudorule/test_sudorules.yml
    tests/sudorule/test_sudorules_member_case_insensitive.yml

### CODE REVIEW

File: tests/sudorule/test_sudorules.yml [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+        state: present
+      register: result
+      failed_when: result.changed or result.failed
+
+    - name: Ensure sudorules can be modified through members
+      ipasudorule:
+        sudorules:
+          - name: testrule1
+            runasuser:
+              - user01
```
Comment (varunmylaraiah) [inline tests/sudorule/test_sudorules.yml:199]: If we look at the next test, 'Ensure sudorules cannot be modified through members, again,' the members user02 in sudorule testrule1 and group02 in sudorule testrule2 are missing here.

```
- name: Ensure sudorules cannot be modified through members, again
      ipasudorule:
        sudorules:
          - name: testrule1
            runasuser:
              - user01
              - user02
          - name: testrule2
            runasuser_group:
              - group01
              - group02
```

**Review by varunmylaraiah — CHANGES_REQUESTED** [review]

**Comment by varunmylaraiah** [comment]
@rjeffman Is it possible to add multiple sudorules using a JSON file?

File: tests/sudorule/test_sudorules.yml (hunk 1 above)
Comment (rjeffman) [reply tests/sudorule/test_sudorules.yml:199]: That's expected. After this playbook both users and groups should be part of of the test rule. The "again" test is to ensure that the values were added to the sudorule, but no value was removed.

**Comment by rjeffman** [comment]
The pydocstyle check is failing due to a false positive, please, see PR #1310.

File: plugins/module_utils/ansible_freeipa_module.py [hunk 2]
```diff
@@ -1584,3 +1588,249 @@ def tm_warn(self, warning):
...
+
+    def check_invalid_parameter_usage(self, entry_dict, fail_on_check=True):
+        """
+        Check if entry_dict parameters are valid for the current state/action.
+
+        Parameters
+        ----------
+            entry_dict: A dictionary representing the module parameters.
+            fail_on_check: If set to True wil make the module execution fail
+                if invalid parameters are used.
```
Comment (t-woerner) [inline plugins/module_utils/ansible_freeipa_module.py:1711]: pydocstyle is not happy with these lines.

**Comment by varunmylaraiah** [comment]
Hi Rafael, I'm seeing an error and not sure what I might be missing here

```
# cat sudo1.yaml 
---
- name: Playbook to ensure a sudorule is present.
  hosts: ipaserver
  become: true

  tasks:
  - name: Test multiple sudo rules
    ipasudorule:
      ipaadmin_password: <xxxxxxx>
      sudorules:
        - name: sudorule1
        - name: sudorule2
        - name: sudorule3
      state: present
```

```
PLAY [Playbook to ensure a sudorule is present.] *******************************************************

TASK [Gathering Facts] *********************************************************************************
task path: /root/sudo1.yaml:2
ok: [master.ipadomain.test]

TASK [Test multiple sudo rules] ************************************************************************
task path: /root/sudo1.yaml:7
fatal: [master.ipadomain.test]: FAILED! => {"changed": false, "msg": "missing required arguments: name"}

PLAY RECAP *********************************************************************************************
master.ipadomain.test      : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
```

File: README-sudorule.md [hunk 3]
```diff
@@ -129,6 +129,24 @@ Example playbook to make sure Sudo Rule is absent:
...
+```yaml
+---
+- name: Playbook to handle sudorules
+  hosts: ipaserver
+  become: true
+
+- name: Ensure multiple Sudo Rules are present using batch mode.
+  ipasudorule:
+    ipaadmin_password: SomeADMINpassword
+    sudorules:
```
Comment (varunmylaraiah) [inline README-sudorule.md:143]: Documentation for the sudorules variable is missing.

File: plugins/modules/ipasudorule.py [hunk 4]
```diff
@@ -232,187 +352,213 @@
     ipaadmin_password: SomeADMINpassword
     name: testrule1
     state: absent
+
+# Ensure multiple Sudo Rules are present using batch mode.
```
Comment (t-woerner) [inline plugins/modules/ipasudorule.py:356]: Please add an example with action: member.

File: plugins/modules/ipasudorule.py [hunk 5]
```diff
@@ -232,187 +352,213 @@
...
-    if len(_result["result"]) > 1:
-        module.fail_json(
-            msg="There is more than one sudorule '%s'" % (name))
-    elif len(_result["result"]) == 1:
-        return _result["result"][0]
-
-    return None
+    try:
+        _result = module.ipa_command("sudorule_show", name, _args)
+    except Exception:  # pylint: disable=broad-except
```
Comment (t-woerner) [inline plugins/modules/ipasudorule.py:383]: Please use ipalib_errors.NotFound here.

File: README-sudorule.md (hunk 3 above)
Comment (rjeffman) [reply README-sudorule.md:143]: Good catch. Fixed.

File: plugins/modules/ipasudorule.py (hunk 5 above)
Comment (rjeffman) [reply plugins/modules/ipasudorule.py]: Fixed.

File: plugins/modules/ipasudorule.py (hunk 4 above)
Comment (rjeffman) [reply plugins/modules/ipasudorule.py:356]: Fixed.

**Review by varunmylaraiah — APPROVED** [review]
The PR looks good to me, and downstream tests have passed with this PR

File: plugins/module_utils/ansible_freeipa_module.py [hunk 6]
```diff
@@ -1584,3 +1588,237 @@ def tm_warn(self, warning):
...
+                "name": {
+                    "aliases": ["cn"]
+                },
+                "description": {}
+                "user": {
+                    "convert": [convert_param_value_to_lowercase]
+                },
+                "group": {
+                    "convert": [convert_param_value_to_lowercase]
+                    "aliases": "ipagroup"
```
Comment (t-woerner) [inline plugins/module_utils/ansible_freeipa_module.py:1649]: Here is a remain of "aliases".

File: plugins/module_utils/ansible_freeipa_module.py [hunk 7]
```diff
@@ -1584,3 +1588,237 @@ def tm_warn(self, warning):
...
+            entry.some_field = mydata
+            return entry
+
+        def main():
+            # ...
+            # Create param mapping, all moudle parameters must be
+            # present as keys of this dictionary
+            params = {
+                "name": {
+                    "aliases": ["cn"]
```
Comment (t-woerner) [inline plugins/module_utils/ansible_freeipa_module.py:1641]: Here is a remain of "aliases".

File: plugins/module_utils/ansible_freeipa_module.py (hunk 6 above)
Comment (rjeffman) [reply plugins/module_utils/ansible_freeipa_module.py]: Fixed

File: plugins/module_utils/ansible_freeipa_module.py (hunk 7 above)
Comment (rjeffman) [reply plugins/module_utils/ansible_freeipa_module.py]: Fixed

File: tests/sudorule/test_sudorule_client_context.yml [hunk 8]
```diff
@@ -37,3 +37,15 @@
...
+  ansible.builtin.import_playbook: test_sudorules.yml
+  when: groups['ipaclients']
+  vars:
+    ipa_test_host: ipaclients
+
+- name: Test sudorule using client context, in server host.
+  ansible.builtin.import_playbook: test_sudorules.yml
+  when: groups['ipaclients'] is not defined or not groups['ipaclients']
+  vars:
+    ipa_context: client
```
Comment (t-woerner) [inline tests/sudorule/test_sudorule_client_context.yml:51]: Why is this here 2 times now?

File: plugins/modules/ipasudorule.py [hunk 9]
```diff
@@ -489,88 +666,100 @@ def main():
...
                     # the provided list against both users and external
                     # users list.
                     runasuser_add, runasuser_del = gen_add_del_lists(
-                        runasuser,
+                        entry.runasuser,
                         (
-                            res_find.get('ipasudorunas_user', [])
-                            + res_find.get('ipasudorunasextuser', [])
+                            list(res_find.get('ipasudorunas_user', []))
+                            + list(res_find.get('ipasudorunasextuser', []))
```
Comment (t-woerner) [inline plugins/modules/ipasudorule.py:757]: In which case is the additional list( .. ) needed?

File: plugins/modules/ipasudorule.py [hunk 10]
```diff
@@ -581,167 +770,165 @@ def main():
                     # the provided list against both groups and external
                     # groups list.
                     runasgroup_add, runasgroup_del = gen_add_del_lists(
-                        runasgroup,
+                        entry.runasgroup,
                         (
-                            res_find.get('ipasudorunasgroup_group', [])
-                            + res_find.get('ipasudorunasextgroup', [])
+                            list(res_find.get('ipasudorunasgroup_group', []))
+                            + list(res_find.get('ipasudorunasextgroup', []))
```
Comment (t-woerner) [inline plugins/modules/ipasudorule.py:776]: In which case is the additional list( .. ) needed?

Comment (t-woerner) [reply plugins/modules/ipasudorule.py:776]: Ok, needed now due to the use of _show.
